internal/expr/prepare.go

```diff
@@ -50,47 +49,54 @@ func getKeys[T any](m map[string]T) []string {
 	return keys
 }
 
-func starCount(fns []fullName) int {
+func starCount[T any](ts []T) int {
 	s := 0
-	for _, fn := range fns {
-		if fn.name == "*" {
-			s++
+	for _, t := range ts {
+		if fn, ok := any(t).(fullName); ok {
+			if fn.name == "*" {
+				s++
+			}
 		}
 	}
 	return s
 }
 
 // prepareInput checks that the input expression corresponds to a known type.
-func prepareInput(ti typeNameToInfo, p *inputPart) (typeMember, error) {
+func (pe *PreparedExpr) prepareInput(ti typeNameToInfo, p *inputPart) error {
 	info, ok := ti[p.sourceType.prefix]
 	if !ok {
 		ts := getKeys(ti)
 		if len(ts) == 0 {
-			return nil, fmt.Errorf(`type %q not passed as a parameter`, p.sourceType.prefix)
+			return fmt.Errorf(`type %q not passed as a parameter`, p.sourceType.prefix)
 		} else {
-			return nil, fmt.Errorf(`type %q not passed as a parameter, have: %s`, p.sourceType.prefix, strings.Join(ts, ", "))
+			return fmt.Errorf(`type %q not passed as a parameter, have: %s`, p.sourceType.prefix, strings.Join(ts, ", "))
 		}
 	}
+	var tm typeMember
 	switch info := info.(type) {
 	case *mapInfo:
-		return &mapKey{name: p.sourceType.name, mapType: info.typ()}, nil
+		tm = &mapKey{name: p.sourceType.name, mapType: info.typ()}
 	case *structInfo:
-		f, ok := info.tagToField[p.sourceType.name]
+		tm, ok = info.tagToField[p.sourceType.name]
 		if !ok {
-			return nil, fmt.Errorf(`type %q has no %q db tag`, info.typ().Name(), p.sourceType.name)
+			return fmt.Errorf(`type %q has no %q db tag`, info.typ().Name(), p.sourceType.name)
 		}
-		return f, nil
 	default:
-		return nil, fmt.Errorf(`internal error: unknown info type: %T`, info)
+		return fmt.Errorf(`internal error: unknown info type: %T`, info)
 	}
+	pe.sql += "@sqlair_" + strconv.Itoa(len(pe.inputs))
+	pe.inputs = append(pe.inputs, tm)
+	return nil
 }
 
 // prepareOutput checks that the output expressions correspond to known types.
 // It then checks they are formatted correctly and finally generates the columns for the query.
-func prepareOutput(ti typeNameToInfo, p *outputPart) ([]fullName, []typeMember, error) {
-	var outCols = make([]fullName, 0)
-	var typeMembers = make([]typeMember, 0)
-
+func (pe *PreparedExpr) prepareOutput(ti typeNameToInfo, p *outputPart) (err error) {
+	defer func() {
+		if err == nil && len(pe.sql) > 1 {
+			pe.sql = pe.sql[:len(pe.sql)-2]
+		}
+	}()
```

Comment:
Can we write a comment in this function saying that this is removing the trailing `, ` that we are adding in the output (which seems to be what is doing)?

Also, this seems to be rather contrived, is the alternative of tracking the state and not outputting the trailing comma much more complex?